_posts/2024-01-21-comparing-cas-and-karpenter.markdown

```diff
@@ -0,0 +1,274 @@
+types based on the workload constraints means that it can request a broad
+range of instances (limited by user configuration) from the infrastructure. By
+contrast, CAS can be configured with many instance types and when presented
+with a decision about which type of instance to request for a specific workload
+it will use the user's preference (up to and including [custom code][cas-grpc])
+to make that choice. The CAS is limited by its node group configurations,
+which are different on each platform, and may or may not support dynamic
+instance discovery and full resource and cost data.
+
+In both cases, CAS and Karpenter, the user has control over the options for
+how nodes are chosen, with preferences and priorities across several ranges
+such as pricing and resource consumption.
+
+## Disrupting behavior and bin packing
+
+Another topic that comes up frequently when talking about Karpenter is
+[bin packing][binpack]. Bin packing in this context refers to the
+algorithms that a program uses to fit items into a set of groups (bins)
+based on arbitrary constraints. Both CAS and Karpenter use bin packing when
+calculating how pods could fit into the possible node choices available. What
+people are most frequently referring to when talking about bin packing and
+Karpenter is its [consolidation and disruption features][disrupt].
+
+A frequent problem with distributed systems that have workloads which come and
+go over time, is that the cluster can become sparsely populated. As workloads
+are removed and not replaced they leave _holes_ in the nodes. This in turn
+causes nodes within a cluster to become underutilized. In
+these situations an activity is required to rebalance the workloads and
+resources in the cluster to optimize usage. CAS and Karpenter both have
+features to help address this problem, with Karpenter providing a more active
+approach.
+
+When using the CAS, users have the ability to configure node resource
+utilization thresholds and inactivity timers. Resource usage is calculated
+based on summed pod resource requests compared to node allocatable capacities.
+When utilization falls below the threshold, a node is cordoned and drained, to
+allow for graceful termination before being removed. The inactivity timer
+provides the mechanism for dictating how long underutilized nodes should
+persist in the cluster.
+
+Karpenter gives users the ability to define several different consolidation
+and disruption conditions under which it can reconfigure the cluster. In the
+example of sparse workloads, Karpenter can consolidate the cluster on a
+regular basis, preferring to choose instances that are cheaper or more resource
+efficient. Karpenter also allows users to replace nodes based on cluster
+configuration skew, age of node in the cluster, and manual intervention. In all
+cases, when Karpenter removes nodes it follows an orderly eviction process to
+allow for graceful node termination.
```

Comment:
I got a little lost in the details in the previous 2 paragraphs and ended up thinking the 2 systems are basically the same when it comes to rebalancing. Is that accurate?

Reply:
no, rebalance is not the right concept to apply across both.

they both have a similar capability to remove nodes that are being underutilized. CAS can only adjust this through the utilization threshold and inactivity timer. Karpenter provides a structured way to create layered rules for when nodes should be removed. CAS wiill not move workloads outside of when it removes a node due to low utilization.

in addition to removing under-utilized nodes, Karpenter also provides an entire suite of options for when to disrupt a cluster and when to perform pod consolidation. these features do not have equivalent in CAS.

i'm thinking i should add some language here to make it clear that karpenter goes above and beyond what CAS can do in this area.

Reply:
i added a few more details to those paragraphs to draw a starker contrast.